Add avatar support for codeberg in git blame

WeetHet opened the pull request:
Release Notes:

- Added support for avatars in git blame for repositories hosted on codeberg

<img width="1144" alt="Screenshot 2024-04-25 at 16 45 22" src="https://github.com/zed-industries/zed/assets/43210583/d44770d8-44ea-4c6b-a1c0-ac2d1d49408f">

Questions:
- Should we move git stuff like `Commit`, `Author`, etc outside of hosting-specific files (I don't think so, as other hostings can have different stuff)
- Should we also add support for self hosted forgejo instances or should it be a different PR?

WeetHet committed:
Add avatar support for codeberg in git blame

WeetHet committed:
Extract author into a different module

- Should we make all author struct fields optional in case some hostings
don't support some of its fields?

WeetHet commented:
I've converted this to a draft as some things have to be decided on first

mrnugget commented on crates/git/src/hosting_provider.rs:
```diff
@@ -71,24 +71,25 @@ impl HostingProvider {
+        Ok(author
+            .map(|author| -> Result<Url, url::ParseError> {
+                let mut url = Url::parse(&author.avatar_url)?;
+                url.set_query(Some("size=128"));
```
Does codeberg even support this?

WeetHet replied:
Size? Probably not, but it doesn't break it, so whatever I guess?

WeetHet replied:
I can add a match inside to only set it for GitHub

WeetHet committed:
Remove unused fields for the Author struct

Let's leave only the avatar_url for now, more fields can be added later
if needed

WeetHet committed:
Only set size=128 for github

WeetHet commented:
> > * Should we move git stuff like `Commit`, `Author`, etc outside of hosting-specific files (I don't think so, as other hostings can have different stuff)
> 
> No, I don't think that's necessary. These are really just there to match GitHub's API schema.
> 
> I think it's fine for now for all the providers to simply return a URL.

I've made it a struct holding only the URL for now, if in the future other stuff will be needed, we can always add new fields

mrnugget commented on Cargo.lock:
```diff
@@ -11095,6 +11095,7 @@ dependencies = [
  "futures 0.3.28",
  "gpui",
  "indoc",
+ "itertools 0.11.0",
```
I don't think you need this. It's fixed on `main`.

WeetHet replied:
Cargo really insists that I do need this, but okay, I've forcefully removed it for now

mrnugget commented on crates/git/src/hosting_provider.rs:
```diff
@@ -71,24 +71,32 @@ impl HostingProvider {
+                (
+                    codeberg::fetch_codeberg_commit_author(repo_owner, repo, &commit, &client)
+                        .await?,
+                    None,
+                )
+            }
+            _ => (None, None),
```
This seems overly complicated: when only 1/3 cases uses the URL query, I think we should set the size inside the `HostingProvider::Github` arm.

WeetHet replied:
Okay, done

WeetHet committed:
Move query to github

WeetHet committed:
Remove itertools from Cargo.lock
